### DISCUSSION TITLE
Testing, dependencies, and the Hyrum’s Law

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#2977] [Ideas]
After reading this quite thought provoking article https://hynek.me/articles/semver-will-not-save-you/ I was wondering what can be improved about things dependency mangement etc. Consider this somewhat random brainstorming for far future. Don't rush. :)

Please read the article first if you have the time. Author there argues that SemVer is nearly meaningless, which I suspect is true. He suggests that all downstream projects should test every time they upgrade upstream dependencies, which I agree with. However I have some ideas about this:

* In principle, every application should have its own test suite *for the dependencies*. Because if it makes an assumption that a dependency does something specific and the assumption isn't tested then you have no guarantee the assumption won't change. We had this issue with `Amount` formatting recently.
* However, if several applications make the same assumption then each having the same test is quite wasteful. That's why having tests in the library is useful. These should be **integration** tests though!
* An issue with tests is that when people want to change behavior they just change the test too. And here comes an idea: let's have a special integration test suite that is forbidden from changing without bumping major version. Then any application can reference those tests when analyzing the code. This should be enforced by CI.
* Allow anyone who makes an assumption to add tests but then do some long hard thinking before accepting it if we want to commit to it.
* As another layer of protection, we could do a crater run - for all public crates depending on our crates we try to run their tests. If we want to break them we can coordinate to make sure breakage is fine for them. The test could run daily in a separate CI since I expect it to be slow. And it should be ran during the release process too.
* Private crates can test on their own and report results.

An extra (crazy?) idea for better version resolution, that would need Cargo cooperation: if the immutable integration tests were standardized one could hash them and refer to the sets of those hashes rather than major version when specifying dependencies. Not only it'd prevent accidental breakage but it'd allow one to depend on a library that broke something unrelated that's not being used.

**Comment by storopoli** [comment]
> Please read the article first if you have the time. Author there argues that SemVer is nearly meaningless, which I suspect is true.

No, it is not meaningless.
The article's context is Python packaging (which is a ~~clusterfuck~~ mess).
You can't yank packager from PIP.
So users that semver pin might be in a insecure version.
Also, with the risk of being labelled toxic, Python developers have a long track-history of screwing up semver.

> - In principle, every application should have its own test suite for the dependencies. Because if it makes an assumption that a dependency does something specific and the assumption isn't tested then you have no guarantee the assumption won't change. We had this issue with Amount formatting recently.

I don't think this is necessary.
Every time a dependency rugpulls us, we learn and add devices (e.g. tests) to prevent this from happening.

> - As another layer of protection, we could do a crater run - for all public crates depending on our crates we try to run their tests. If we want to break them we can coordinate to make sure breakage is fine for them. The test could run daily in a separate CI since I expect it to be slow. And it should be ran during the release process too.

I don't think this is viable.
It would shift a lot of resources towards that and we don't have the luxury.

---

## My two cents

I think that we need to be extra diligent when adding dependencies.
Take for example `bitcoin`, the only dependencies that are outside of our control,
i.e. none of the maintainers are "regulars" here,
are `base64` and `serde` (I know I am not listing transitive dependencies).

I don't think we need to explicitly test dependencies like you are proposing.

Breakages will happen and we need to keep an antifragile approach towards them.
Every time that we get rugpulled by a dependency, we make sure that such cornercase won't happen again.

**Reply by apoelstra** [reply]
I'm not sure a crater run would be so bad. Looking at crates.io it seems we have <400 direct dependencies in bitcoin. I'd be surprised if the total number of transitive dependencies were more than 1000, and it's definitely not 10000. I was doing something with crates-index recently but I can't find the code now .. but it's not too hard to pull a complete list.

>I think that we need to be extra diligent when adding dependencies.

Lol, more diligent than we are now? Right now `cc` `base64`, `serde` and `rand` are the only dependencies in rust-bitcoin that we don't personally maintain (there are a couple more dev-dependencies). And we have pinned every one of those at times, maintained a fork of base64 for a long time, left `rand` at some ancient major version, and have an ongoing project to eliminate the proc-macros from `serde`.

**Reply by Kixunil** [reply]
> No, it is not meaningless.

I didn't say that. Though your insight on Python being just worse in this regard is useful. Thank you for that!

Still, even in Rust we're not perfect and it's useful to try to think what we can do better.

> Every time a dependency rugpulls us, we learn and add devices (e.g. tests) to prevent this from happening.

We might make some assumptions about other dependency code and we're a security-critical library. It makes sense to be more proactive.

> It would shift a lot of resources towards that and we don't have the luxury.

Really? How much computing power do we need per day? We could also just reduce the frequency to weekly or worst case, do it before RC releases (we just need to give ourselves enough time to fix stuff). This is not a question of talent - any script kiddie could write the code to run it. There probably already is a tool for this (I think I've seen it). So it's just about money. There's enough companies using the library that we should be able to get some funding here. (And I don't think it'll be a lot anyway.)

> Lol, more diligent than we are now? 

Exactly. In some cases I'd prefer if we could actually relax a bit since more code sharing also brings less (security) bugs due to having more reviewers.

**Comment by apoelstra** [comment]
I'm certainly in favor of integration testing. I think in practice we'll wind up adding things as we encounter them, which is unfortunate but at least simplifies the "do we want to commit to this?" question -- since if we broke something and regret it, then presumably we want to commit to it.

Having integration tests for all our macros would be a great start, since it's impossible to test macros "in-crate" and we've caused breakage repeatedly because of this.

FWIW in the case of address formatting a unit test would have been sufficient.

**Reply by Kixunil** [reply]
I think there's a bunch of cases where we want to commit to things. Many of them are used in the ecosystem.

I'm not sure what address formatting you're referring to. Do you mean amount? I think testing an API in an integration test is a stronger commitment, especially if that test is in a file explicitly marked "do not change".

**Reply by apoelstra** [reply]
Yeah, I meant to say amount.

**Comment by tcharding** [comment]
> How do I know when to release 1.0.0?

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backward compatibility, you should probably already be 1.0.0.

This text from the blog post hit me.

> we could do a crater run

What does "crater run" mean please? Is this a Rust term for "check and/or do something to all crates that depend on another crate"?

>  for all public crates depending on our crates we try to run their tests.

Do you mean we maintain a live branch of dependent projects that tracks `master` of `bitcoin` so that we can run their tests after we change `bitcoin`? If so I've thought about this but it sounds painful. That thought process is what led me to wanting to do draft PRs for our RC releases on dependant crates but I only ever managed to do `miniscript` and `bdk`'s stack of crates.


My two cents:

In the semver-check PR discussion recently the `cargo-semver-checks` author claimed that "any observable functionality is part of the public API" - this made me think already that semver is total smoke and mirrors.

An idea I had was that we should just have two discreet explicit modes for our crates 

1. break as much shit as we want
2.  try not to break anything

As for releases, also two modes (both assuming we have already released 1.0.0) :

1. Release versions very often (eg, we can look at the three commits we are releasing and can be confident what has changed) 
5. Release major version every X months.

**Reply by Kixunil** [reply]
> This text from the blog post hit me.

Yes, we're not strictly adhering to it. But let's just continue - it'll be over somewhat soon.

> What does "crater run" mean please?

"crater" tests all the crates on crates.io with new version of Rust or when there's a PR that is likely to break something they the PR. (They also compare if the crate builds on previous version to exclude always-failing tests. Which reminds me, us not publishing test data prevents us from participating.) Obviously in our context it refers to us testing all our dependent crates.

> Do you mean we maintain a live branch of dependent projects

No maintaining. Just have an automated script to pull the released version and test it against the PR.

> author claimed that "any observable functionality is part of the public API" - this made me think already that semver is total smoke and mirrors.

Yeah, somewhat. It's really hard to get right if even possible and then if you really need to break something that almost nobody uses you're in trouble because bumping major version will cause libcpocalypse, not bumping it is strictly wrong and not changing it is also bad.

I don't mind doing minor releases more often once we hit stable. But I think it's not really about the number of commits but about the tests.

> Release major version every X months.

If for the base crates X = ~200 then I'm fine with it.

**Reply by tcharding** [reply]
> No maintaining. Just have an automated script to pull the released version and test it against the PR.

Pull the released version of a downstream crate? Downstream crates won't build using `rust-bitcoin` master because we break so much stuff.   Or am I being brain dead?

**Reply by Kixunil** [reply]
You know that `[patch]` table exists in `Cargo.toml`, right? :) Note that I'm only speaking about the stable crates.

**Reply by storopoli** [reply]
Oh I see the "crater run" is just a `cargo test` in all deps?
That is not that hard to do. I think it's an interesting idea.

**Reply by Kixunil** [reply]
Almost. We need to first run `cargo test` to check their tests are not failing for other reasons (in which case we log it but consider it success), then patch the dep to use `master` then run `cargo test` again and report results.

**Reply by tcharding** [reply]
> Note that I'm only speaking about the stable crates.

That is more reasonable, I thought you meant for `rust-bitcoin`.

**Reply by tcharding** [reply]
Excuse me for being autistic but I'm still confused, can I clarify some terms:

- "crater run": Is this referring to the [crater](https://github.com/rust-lang/crater) tool or is it a term that means "run through some set of crates and do some stuff for each crate"?
- "deps": We are talking about dependents of `rust-bitcoin` right, not dependencies?

**Reply by apoelstra** [reply]
I have interpreted it as "run through the set of crates that are dependent on rust-bitcoin and `cargo test` them".

Though I do wonder if the actual crater tool can do this for us..

**Reply by tcharding** [reply]
In case its lost in text, I wrote this whole message with a smile on my face.

> crates that are dependent on rust-bitcoin

ooph, grammatically correct but confusing still because of the terminology used on `crates.io`. Natural language, what a PITA.

I copied terminology from crates.io:

- dependencies = upstream
- dependents = downstream

Aaaagh, coooool, I wonder if this language problem is why we have the terms upstream and downstream to start with ???

We are talking about running tests in crates _downstream_ from `rust-bitcoin`.

And now back to my original argument, this is a great idea and will be a massive punish because it requires us maintaining a patched version of each downstream crate that tracks `rust-bitcoin` master, bags not being the muppet who has to rebase that constantly. (And all the manifest patching involved, yes @Kixunil I am acutely aware that `[patch]` exists, while useful it has stolen more than a few hours of my life :)

**Reply by Kixunil** [reply]
What @apoelstra says except not really the `bitcoin` crate since those would start failing really soon. I meant `units` and primitives once they are released and people start depending on them directly. Later keys and other stable stuff. I don't think crater itself can do it but I believe I've seen something similar. It shouldn't be too hard to make anyway.

> I wonder if this language problem is why we have the terms upstream and downstream to start with ???

I believe it is.

> it requires us maintaining a patched version of each downstream crate that tracks `rust-bitcoin` master

It doesn't. We download the crate from crates.io, run its tests, if they succeed append the patch to `Cargo.toml` run tests again and report the results.

> And all the manifest patching involved

It's literally `echo -e '[patch.crates-io]\nbitcoin-units = { git = "..." }' >> Cargo.toml` What do you want to maintain about that?

**Reply by tcharding** [reply]
if this is only for when folks start depending directly on `primitives`, `units` etc. then got for it! I just don't want to do it for `bitcoin`.

**Reply by Kixunil** [reply]
> I just don't want to do it for `bitcoin`.

I'm not insane enough to propose that. At least not until `bitcoin` itself is stable in 200 years.

**Comment by dr-orlovsky** [comment]
Good that I did [Strict Types](hytps://strict-types.org) which allow me to abstract from semver and use formal verification of any breaking changes un data structures and their serialization - and clearly see which specific part has broken.

An example of how you compile rust data types i to strict types can be found for instance in [BP Core library](https://github.com/BP-WG/bp-core) - a modular and modern alternative to `rust-bitcoin` consensus part - or in [RGB](https://github.com/RGB-WG/rgb-core). Check the `stl` dirs there to see auto-compiled strict types descriptions with semantic hashes coming from `StrictType` derive macros on the structures.

**Reply by Kixunil** [reply]
I don't see how it helps. SemVer verification is essentially Halting problem.

**Reply by dr-orlovsky** [reply]
Instead of relying on SemVer one should be using formal system verifying binary and semantic compatibility. Strict types provide such system. Both serialization/memory layout and semantics create a unique hash id, and once it changes you know about the breaking change - and can narrow down where it had happened. Thus, you depend on libraries providing specific semantic hash and do not update if it changed.

**Reply by Kixunil** [reply]
Does it mean you consider all changes breaking?

**Reply by dr-orlovsky** [reply]
No, it doesn't.

There might be changes where you have semantic change but no memory layout change (change in field or type names). The seveloper has an option to decide whether to tolerate such changes.

Also, there might be non-breaking changes, like some cases of new-types wrappings.

**Reply by Kixunil** [reply]
Sounds crazy to be possible. I'll definitely need to take a closer look.
